payment_ippay/models/payment.py

```diff
@@ -0,0 +1,156 @@
+                # and check if there is two components containing only digits
+                # this fixes multiples crashes, if there was no space
+                # between the '/' and the components the code was crashing
+                # the code was also crashing if the customer
+                # was proving non digits to the date.
+                cc_expiry = [i.strip() for i in data["cc_expiry"].split("/")]
+                all_digits = all(i.isdigit() for i in cc_expiry)
+                if len(cc_expiry) != 2 or not all_digits:
+                    return False
+                try:
+                    if datetime.now().strftime("%y%m") > datetime.strptime(
+                        "/".join(cc_expiry), "%m/%y"
+                    ).strftime("%y%m"):
+                        return False
+                except ValueError:
+                    return False
+            return False if error else True
+        return True
+
+
+class PaymentToken(models.Model):
+    _inherit = "payment.token"
+
+    @api.model
+    def _ippay_get_token(self, values):
+        acquirer = self.env["payment.acquirer"].browse(values["acquirer_id"])
+        expiry = (values["cc_expiry"]).split("/")
+        if values.get("cc_number"):
+            values["cc_number"] = values["cc_number"].replace(" ", "")
+            card_detail = {
+                "cc_number": values["cc_number"],
+                "expiry_month": values.get("cc_expiry_month")
+                or expiry[0].replace(" ", ""),
+                "expiry_year": values.get("cc_expiry_year")
+                or expiry[1].replace(" ", ""),
+            }
+
+            xml = """<ippay>
+            <TransactionType>TOKENIZE</TransactionType>
+            <TerminalID>%s</TerminalID>
+            <CardNum>%s</CardNum>
+            <CardExpMonth>%s</CardExpMonth>
+            <CardExpYear>%s</CardExpYear>
+            </ippay>""" % (
+                acquirer.ippay_terminal_id,
+                card_detail.get("cc_number"),
+                card_detail.get("expiry_month"),
+                card_detail.get("expiry_year"),
+            )
+            if acquirer.api_url:
+                url = acquirer.api_url
+            r = requests.post(url, data=xml, headers={"Content-Type": "text/xml"})
+            data = xmltodict.parse(r.content)
+            token = data["IPPayResponse"].get("Token")
+            if not token:
+                raise ValidationError(
+                    _("Customer payment token creation in IPpay failed: %s")
+                    % (data["IPPayResponse"].get("ErrMsg"))
+                )
+            else:
+                return token
+
+    @api.model
+    def ippay_create(self, values, token_code=False):
+        # In case we already know the token assigned, just use it
+        token_code = token_code or self._ippay_get_token(values)
+        existing = self.sudo().search(
+            [
+                ("acquirer_ref", "=", token_code)(
```

Comment:
@dreispt :
Looks like missing ',' after tuple ("acquirer_ref", "=", token_code)
Should be ("acquirer_ref", "=", token_code), ("partner_id"*******)